File: TP/CH04/morphosyntaxe/java/MorphoSyntaxe.java
import java.util.regex.*;
import java.io.*;
import java.util.*;
import java.lang.Math;
import java.util.stream.*;
import java.util.Arrays;

public class MorphoSyntaxe {
	public static enum OPT {FORCE_BRUTE, VITERBI, GOURMAND}
	private HMM modele;
	private List<Tuple<List<String>, List<String>>> eval = new ArrayList<>();

	public void entrainer(String url){
		try {
			BufferedReader f = new BufferedReader(new FileReader(url));
			List<List<Tuple<String, String>>> data = new ArrayList<>();
			Set<String> tagsList = new LinkedHashSet<String>();
			String l;
			while((l = f.readLine()) != null) {
				String[] phrase = l.trim().toLowerCase().replace("\\/", "@@").split(" ");
				if (phrase.length > 0){
					List<Tuple<String, String>> p = new ArrayList<>();
					for (String mottag: phrase){
						String[] mt = mottag.split("/");
						p.add(new Tuple<String, String>(mt[0].replace("@@", "/"), mt[1]));
						tagsList.add(mt[1]);
					}
					data.add(p);
				}
			}
			this.modele = new HMM(new ArrayList<String>(tagsList));
			this.modele.entrainer(data);
			f.close();
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	private List<String> estimer(List<String> phrase, OPT opt){
		List<String> resTags;
		switch(opt){
			case VITERBI:
			resTags = this.modele.estimerViterbi(phrase);
			break;
			case GOURMAND:
			resTags = this.modele.estimerGourmand(phrase);
			break;
			default:
			resTags = this.modele.estimerForceBrute(phrase);
		}
		return resTags;

		// return IntStream.range(0, resTags.size())
		// 	.mapToObj(i -> new Tuple<String, String>(phrase[i], resTags.get(i)))
		// 	.collect(Collectors.toList());
	}

	public List<String> estimer(String phrase, OPT opt){
		return estimer(Arrays.asList(phrase.trim().toLowerCase().split(" ")), opt);
	}

	public void chargerModele(String url){
		try {
			FileInputStream f = new FileInputStream(url);
			ObjectInputStream in = new ObjectInputStream(f);
			modele = (HMM) in.readObject();
			in.close();
			f.close();
		} catch(IOException|ClassNotFoundException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void sauvegarderModele(String url){
		try {
			FileOutputStream f = new FileOutputStream(url);
			ObjectOutputStream out = new ObjectOutputStream(f);
			out.writeObject(modele);
			out.close();
			f.close();
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void chargerEvaluation(String url){
		try {
			BufferedReader f = new BufferedReader(new FileReader(url));
			String l;
			while((l = f.readLine()) != null) {
				String[] mottags = l.trim().toLowerCase().replace("\\/", "@@").split(" ");
				if (mottags.length > 1){
					List<String> mots = new ArrayList<String>();
					List<String> tags = new ArrayList<String>();
					for (String mottag: mottags){
						String[] mt = mottag.split("/");
						mots.add(mt[0].replace("@@", "/"));
						tags.add(mt[1]);
					}
					this.eval.add(new Tuple<List<String>, List<String>>(mots, tags));
				}
			}
			f.close();
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public double evaluerPhrase(List<String> sysTags, List<String> refTags){
		double INT = 0.0;
		for (int i=0; i< sysTags.size(); i++){
			if (sysTags.get(i).equals(refTags.get(i))) INT++;
		}
		return INT/refTags.size();
	}

	public void evaluer(int n){
		if (n==-1) n = this.eval.size();
		else Collections.shuffle(this.eval);
		double scoreGourmand = 0.0;
        double scoreViterbi = 0.0;
        double scoreForceBrute = 0.0;
		Tuple<List<String>, List<String>> test;
		List<String> sysTags;
		double score;
		for (int i=0; i<n; i++){
			test = this.eval.get(i);
			System.out.println("=======================");
			System.out.println("phrase : " + test.k);
			System.out.println("tags de référence : " + test.v);

			// sysTags = estimer(test.k, OPT.FORCE_BRUTE);
			// System.out.println("brute force : " + sysTags);
			// score = evaluerPhrase(sysTags, test.v);
			// System.out.println("score : " + score);
			// scoreForceBrute += score;

			sysTags = estimer(test.k, OPT.VITERBI);
			System.out.println("viterbi : " + sysTags);
			score = evaluerPhrase(sysTags, test.v);
			System.out.println("score : " + score);
			scoreViterbi += score;

			sysTags = estimer(test.k, OPT.GOURMAND);
			System.out.println("gourmand : " + sysTags);
			score = evaluerPhrase(sysTags, test.v);
			System.out.println("score : " + score);
			scoreGourmand += score;
		}
		scoreGourmand /= n;
		scoreViterbi /= n;
		scoreForceBrute /= n;
		System.out.println("----------- Scores totaux -------------");
		System.out.println("score gourmand : " + scoreGourmand);
		System.out.println("score viterbi : " + scoreViterbi);
		System.out.println("score force brute : " + scoreForceBrute);
	}


	// TESTS
	//=======================
	public static void tester_noter(){
		MorphoSyntaxe annotateur = new MorphoSyntaxe();
		annotateur.entrainer("./data/test.txt");
		String phrase = "il peut aider";
		
		String[] tags = new String[]{"det", "verb", "verb"};
		double[] scores = new double[]{-3.465735902799726, -3.401197381662155, -3.3141860046725258};
		String[] mots = phrase.toLowerCase().split(" ");
		String past_tag="";
		for (int i=0; i<mots.length; i++){
			System.out.print("mot=" + mots[i] + ", tag=" + tags[i] + ", score=" + scores[i]);
			System.out.println(", mon score =" + annotateur.modele.noter(past_tag, tags[i], mots[i]));
			past_tag = tags[i];
		}

	}

	public static void tester_viterbi(){
		MorphoSyntaxe annotateur = new MorphoSyntaxe();
		annotateur.entrainer("./data/univ_train.txt");
		String phrase = "this has some logic .";
		
		System.out.println("viterbi doit être comme force brute");
		System.out.println("force brute : " + annotateur.estimer(phrase, OPT.FORCE_BRUTE));
		System.out.println("gourmand : " + annotateur.estimer(phrase, OPT.GOURMAND));
		System.out.println("viterbi : " + annotateur.estimer(phrase, OPT.VITERBI));
	}

	public static void evaluation(){
		MorphoSyntaxe annotateur = new MorphoSyntaxe();
		annotateur.entrainer("./data/univ_train.txt");
		annotateur.chargerEvaluation("./data/univ_test.txt");
		annotateur.evaluer(-1);
	}

	public static void main(String[] args) {
		tester_noter();
		//tester_viterbi();
		//evaluation();
	}
}

class HMM implements Serializable {
	// Tag_i --> tag_i+1 --> prob // probabilités de transition
	private Map<String, Map<String, Double>> transition = new HashMap<>();
	// Tag --> mot --> prob // probabilité d'émission
	private Map<String, Map<String, Double>> emission = new HashMap<>();
	// Tag --> prob // probabilité initiale des états
	private Map<String, Double> init = new HashMap<>();
	private List<String> tags;
	// score des mots non trouvées dans l'emission
	// unk_score = 1/|V|
	private double unkScore = 0.0;

	public void printEmission(){
		System.out.println(emission);
	}

	public HMM(List<String> tags){
		this.tags = tags;
		for (String tag: tags){
			transition.put(tag, new HashMap<String, Double>());
			emission.put(tag, new HashMap<String, Double>());
			init.put(tag, 0.0);
			for (String tagSuiv: tags){
				this.transition.get(tag).put(tagSuiv, 0.0);
			}
		}
	}

	public static void initInc(Map<String, Double> map, String cle, double ...val){
		double initVal = val.length > 0 ? val[0] : 1.0;
		double incVal = val.length > 1 ? val[1] : 1.0;
		if (map.containsKey(cle)) map.put(cle, map.get(cle)+ incVal);
		else map.put(cle, initVal);
	}

	public void entrainer(List<List<Tuple<String, String>>> data){
		Map<String, Double> tagsFreq = new HashMap<>();
		Set<String> vocab = new HashSet<>();

		for (List<Tuple<String, String>> phrase : data){
			Tuple<String, String> mottagPrec = phrase.get(0);
			vocab.add(mottagPrec.k);
			initInc(this.init, mottagPrec.v);
			initInc(tagsFreq, mottagPrec.v);
			if (! this.emission.containsKey(mottagPrec.v)){
				this.emission.put(mottagPrec.v, new HashMap<String, Double>());
			}
			initInc(this.emission.get(mottagPrec.v), mottagPrec.k);
			for (int j=1; j< phrase.size(); j++){
				Tuple<String, String> mottag = phrase.get(j);
				vocab.add(mottag.k);
				// transition tag_prec --> tag
				initInc(this.transition.get(mottagPrec.v), mottag.v);
				// émission tag --> mot
				initInc(this.emission.get(mottag.v), mottag.k);
				initInc(tagsFreq, mottag.v);
 				mottagPrec = mottag;
			}
		}

		int V = vocab.size();
		this.unkScore = 1.0/V;
		int Vt = this.tags.size();
		int nbrPhrases = data.size();
		for (String tag: this.tags){
			this.init.put(tag, (this.init.get(tag)+1)/(nbrPhrases+Vt));
			double nbrTag = tagsFreq.get(tag);
			Map<String, Double> emissionTag = this.emission.get(tag);
			for (String mot: emissionTag.keySet()){
				emissionTag.put(mot, (emissionTag.get(mot)+1.0)/(nbrTag + V));
			}
			Map<String, Double> transitionTag = this.transition.get(tag);
			for (String tagSuiv: this.tags){
				transitionTag.put(tagSuiv, (transitionTag.get(tagSuiv)+1.0)/(nbrTag+Vt));
			}
		}
		//////////////////////
	}//entrainer

	//TODO: compléter cette méthode
	public double noter(String pastPos, String currentPos, String mot){
		return Math.log(1);
	}

	public List<String> estimerForceBrute(List<String> phrase){
		List<String> tagsMax = new ArrayList<String>();
		double scoreMax = Double.NEGATIVE_INFINITY;
		LinkedList<Tuple<String,Tuple<Integer,Double>>> pile = new LinkedList<>();
		pile.addLast(new Tuple<String,Tuple<Integer,Double>>("", new Tuple<Integer,Double>(0, 0.0)));
		int T = phrase.size();
		int N = tags.size();
		while (pile.size() > 0){
			int i = pile.size() - 1;
			Tuple<String,Tuple<Integer,Double>> triple = pile.pollLast();
			String pastPos = triple.k;
			int t = triple.v.k;
			double score = triple.v.v;
			pile.addLast(new Tuple<String,Tuple<Integer,Double>>(pastPos, new Tuple<Integer,Double>(t + 1, score)));

			//forward
			while (i < T){
				String currentPos = tags.get(t);
				String mot = phrase.get(i);
				score += noter(pastPos, currentPos, mot);
				pile.addLast(new Tuple<String,Tuple<Integer,Double>>(currentPos, new Tuple<Integer,Double>(1, score)));
				i++;
				t = 0;
				pastPos = currentPos;
			}

			if (scoreMax < score){
				scoreMax = score;
				tagsMax = pile.stream().map(e -> e.k).collect(Collectors.toList());
			}

			//backward
			pile.pollLast(); //dépiler le dernier élément
			while ((pile.size() > 0) && (pile.getLast().v.k == N)) pile.pollLast();
		}
		
		tagsMax.remove(0);

		return tagsMax;
	}

	public List<String> estimerGourmand(List<String> phrase){
		List<String> tags = new ArrayList<String>();
		String pastTag = "";
		for (String mot: phrase){
			double maxScore = Double.NEGATIVE_INFINITY;
			String maxTag = "";
			for (String tag: this.tags){
				double score = this.noter(pastTag, tag, mot);
				if (score > maxScore){
					maxScore = score;
					maxTag = tag;
				}
			}
			tags.add(maxTag);
			pastTag = maxTag;
		}
		return tags;
	}

	//TODO: compléter cette méthode : vous devez suivre la même structure définie ici
	public List<String> estimerViterbi(List<String> phrase){
		int N = tags.size();
		int T = phrase.size();
		double [][] viterbi = new double[N][T];
		int [][] backpointer = new int[N][T];
		

		List<String> resTags = new ArrayList<String>();

		
		return resTags;
	}

}

class Tuple<K, V> {
	public final K k;
	public final V v;
	public Tuple(K k, V v) {
		this.k = k;
		this.v = v;
	}

	public String toString(){
		return "(" + k + ", " + v + ")";
	}
}

File: TP/CH04/morphosyntaxe/python/morphosyntaxe.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import math
import json
import random
from typing import Dict, List, Tuple, Set

def init_inc(mp, cle, init_val=1.0, inc_val=1.0):
    if cle in mp:
        mp[cle] += inc_val
    else:
        mp[cle] = init_val

class HMM:

    def __init__(self, tags: List[str]):
        """_summary_

        :param tags: _description_
        :type tags: List[str]
        """

        # Tag_i --> tag_i+1 --> prob
        self.transition : Dict[str, Dict[str, float]] = {} # probabilités de transition
        # Tag --> mot --> prob
        self.emission : Dict[str, Dict[str, float]] = {} # probabilité d'émission
        # Tag --> prob
        self.init : Dict[str, float] = {} # probabilité initiale des états

        # score des mots non trouvées dans l'emission
        # unk_score = 1/|V|
        self.unk_emission = 0.0

        self.tags = tags

        for tag in tags:
            self.transition[tag] = {}
            self.emission[tag] = {}
            self.init[tag] = 0.0
            for tag_suiv in tags:
                self.transition[tag][tag_suiv] = 0.0

    def entrainer(self, data : List[List[Tuple[str, str]]]) -> None :

        tags_freq = {}
        vocab = set()
        for phrase in data:
            mot_prec, tag_prec = phrase[0]
            vocab.add(mot_prec)
            init_inc(self.init, tag_prec)
            init_inc(tags_freq, tag_prec)
            init_inc(self.emission[tag_prec], mot_prec)

            for j in range(1, len(phrase)):
                mot, tag = phrase[j]
                vocab.add(mot)
                # transition tag_prec --> tag
                init_inc(self.transition[tag_prec], tag)

                # émission tag --> mot
                init_inc(self.emission[tag], mot)

                init_inc(tags_freq, tag)

                mot_prec, tag_prec = mot, tag

        V = len(vocab)
        self.unk_emission = 1/V
        Vt = len(self.tags)
        nbr_phrases = len(data)
        for tag in self.tags:
            self.init[tag] = (self.init[tag]+1)/(nbr_phrases+Vt)
            nbr_tag = tags_freq[tag]
            emission_tag = self.emission[tag]
            for mot in emission_tag :
                emission_tag[mot] = (emission_tag[mot]+1)/(nbr_tag + V)
            transition_tag = self.transition[tag]
            for tag_suiv in self.tags:
                transition_tag[tag_suiv] = (transition_tag[tag_suiv]+1)/(nbr_tag+Vt)

    # TODO: compléter cette méthode
    def noter(self, past_pos: str, current_pos: str, mot: str) -> float:
        return math.log(1)

    def estimer_force_brute(self, phrase: List[str]) -> List[str]:
        tags_max = []
        score_max = - math.inf
        pile = [('', 0, 0.0)]
        T = len(phrase)
        N = len(self.tags)
        while len(pile) > 0 :
            i = len(pile) - 1
            past_tag, t, score = pile.pop()
            pile.append((past_tag, t+1, score))
            # forward
            while i < T:
                tag = self.tags[t]
                mot = phrase[i]
                score += self.noter(past_tag, tag, mot)
                pile.append((tag, 1, score))

                i += 1
                t = 0
                past_tag = tag
            #print(pile)
            if score_max < score:
                #print(pile)
                score_max = score
                tags_max, _, _ = zip(*pile[1:])
                #print(score_max)
                
            # backward
            pile.pop() # dépiler le dernier élément
            while len(pile) > 0 and pile[-1][1] == N:
                pile.pop()

        return list(tags_max)

    def estimer_gourmand(self, phrase: List[str]) -> List[str]:
        tags = []
        past_tag = ''
        for mot in phrase:
            max_score = - math.inf
            max_tag = ''
            for tag in self.tags:
                score = self.noter(past_tag, tag, mot)
                if score > max_score:
                    max_score = score
                    max_tag = tag
            tags.append(max_tag)
            past_tag = max_tag
        return tags

    # TODO compléter cette méthode : vous devez suivre la même structure définie ici
    def estimer_viterbi(self, phrase: List[str]) -> List[str] :

        N = len(self.tags)
        T = len(phrase)
        viterbi = [[-math.inf]*T for i in range(N)]
        backpointer = [[0]*T for i in range(N)]

        tags = []
        
        return tags

    def exporter_json(self):
        return self.__dict__.copy()

    def importer_json(self, data):
        for cle in data:
            self.__dict__[cle] = data[cle]

class MorphoSyntaxe:
    def __init__(self):
        self.eval = []

    @staticmethod
    def lire_phrases_etiquetees(url: str) -> Tuple[List[str], List[str]] :
        
        f = open(url, 'r')
        data = []
        tags = set()
        for l in f: # la lecture ligne par ligne
            if len(l) < 5 :# si la longueur est 
                continue
            # on supprime les espaces auteur de la ligne, on convertit les caractères een minuscule, 
            # on remplace 
            phrase = l.strip().lower().replace('\/', '\\').split()
            p = []
            data.append(p)
            for mottag in phrase:
                mot, tag = mottag.split("/")
                p.append((mot.replace('\\', '/'), tag))
                tags.add(tag)
        f.close()
        return data, list(tags)

    def entrainer(self, url: str) -> None:
        data, tags = self.__class__.lire_phrases_etiquetees(url)

        self.modele = HMM(tags)
        self.modele.entrainer(data)

    def _estimer(self, phrase: List[str], opt: str = 'forcebrute') -> List[Tuple[str, str]]:
        if opt == 'viterbi':
            tags = self.modele.estimer_viterbi(phrase)
        elif opt == 'gourmand':
            tags = self.modele.estimer_gourmand(phrase)
        else :
            tags = self.modele.estimer_force_brute(phrase)
        # res = []
        # for i in range(len(phrase)):
        #     res.append((phrase[i], tags[i]))
        return tags
        #return list(zip(phrase, tags))

    def estimer(self, phrase: str, opt: str = 'bruteforce') -> List[Tuple[str, str]]:
        return self._estimer(phrase.strip().lower().split(), opt=opt)

    def charger_modele(self, url):
        f = open(url, 'r')
        data = json.load(f)
        self.modele.importer_json(data)
        f.close()

    def sauvegarder_modele(self, url):
        f = open(url, 'w')
        json.dump(self.modele.exporter_json(), f)
        f.close()

    def charger_evaluation(self, url):
        senttags, _ = self.__class__.lire_phrases_etiquetees(url)
        for senttag in senttags:
            sents, tags = zip(*senttag)
            self.eval.append((list(sents), list(tags)))
        #self.eval = [(zip(*senttag)) for senttag in senttags]

    def evaluer_phrase(self, sysTags, refTags):
        INT = 0.0
        for i in range(len(sysTags)):
            if sysTags[i] == refTags[i]:
                INT += 1.0
        return INT/len(refTags)

    def evaluer(self, n): #Moyennes des R, P, F1
        if n == -1:
            S = self.eval
            n = len(S)
        else :
            S = random.sample(self.eval, n)
        score_gourmand = 0.0
        score_viterbi = 0.0
        score_force_brute = 0.0
        for i in range(n):
            test = S[i]
            print('=======================')
            print('phrase :', test[0])
            print('tags de référence :', test[1])

            # sysTags = self._estimer(test[0])
            # print('brute force :', sysTags)
            # score_i = self.evaluer_phrase(sysTags, test[1])
            # print('score :', score_i)
            # score_force_brute += score_i

            sysTags = self._estimer(test[0], opt="viterbi")
            print('viterbi :', sysTags)
            score_i = self.evaluer_phrase(sysTags, test[1])
            print('score :', score_i)
            score_viterbi += score_i

            sysTags = self._estimer(test[0], opt="gourmand")
            print('gourmand :', sysTags)
            score_i = self.evaluer_phrase(sysTags, test[1])
            print('score :', score_i)
            score_gourmand += score_i

        score_gourmand = score_gourmand/n
        score_viterbi = score_viterbi/n
        score_force_brute /= n
        print('----------- Scores totaux -------------')
        print('score gourmand : ', score_gourmand)
        print('score viterbi : ', score_viterbi)
        print('score force brute : ', score_force_brute)


def tester_noter():
    annotateur = MorphoSyntaxe()
    annotateur.entrainer('./data/test.txt')
    phrase = 'il peut aider'
    tags = ['det', 'verb', 'verb']
    scores = [-3.465735902799726, -3.401197381662155, -3.3141860046725258]
    mots = phrase.strip().lower().split()
    past_tag=''
    for mot, tag, score in zip(mots, tags, scores):
        print('mot=', mot, ', tag=',tag, ', score=', score, ', mon score =', annotateur.modele.noter(past_tag, tag, mot))
        past_tag=tag

def tester_viterbi():
    annotateur = MorphoSyntaxe()
    annotateur.entrainer('./data/univ_train.txt')
    phrase = 'this has some logic .'
    print('viterbi doit être comme force brute')
    print('force brute : ', annotateur.estimer(phrase))
    print('gourmand : ', annotateur.estimer(phrase, opt='gourmand'))
    print('viterbi : ', annotateur.estimer(phrase, opt='viterbi'))


def evaluation():
    annotateur = MorphoSyntaxe()
    annotateur.entrainer('./data/univ_train.txt')
    annotateur.charger_evaluation('./data/univ_test.txt')
    annotateur.evaluer(-1)


if __name__ == '__main__':
    tester_noter()
    # tester_viterbi()
    # evaluation()
